homework/hw2-goals-approaches/ez/hw2-ez.md

```diff
@@ -0,0 +1,390 @@
+    - Describe: what are the models succeeding at? Where are they struggling? How do you propose addressing these weaknesses next time?
+
++++
+
+### Multiclass
+
+```{code-cell} ipython3
+import pandas as pd
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn import preprocessing
+from sklearn.model_selection import train_test_split
+from sklearn.svm import LinearSVC
+import pickle
+import matplotlib.pyplot as plt
+from sklearn.metrics import plot_confusion_matrix
+```
+
+```{code-cell} ipython3
+# Load the multiclass model
+multiclass_model = pickle.load(open("multiclass.sav", 'rb'))
+```
+
+Preprocessing:
+- Remove NAs in column _flavor_text_, _text_, and _color_identity_
+- Remove multicolored _color_identity_
+- min_df = 5, ignore rare words (appear in less than 5 documents)
+- max_df = 0.8, ignore common words (appear in more than 80% of documents)
+
+```{code-cell} ipython3
+# Load the data
+df = pd.read_feather("../../../data/mtg.feather")
+
+# Remove NAs
+df = df.dropna(subset = ["flavor_text", "text", "color_identity"]).reset_index(drop=True)
+
+# Remove multicolored cards
+df = df[df.color_identity.map(len) < 2]
+```
+
+```{code-cell} ipython3
+# X
+tfidf = TfidfVectorizer(
+    min_df=5, # ignore rare words (appear in less than 5 documents)
+    max_df=0.8, # ignore common words (appear in more than 80% of documents)
+    stop_words="english"
+)
+text = df["text"] + df["flavor_text"].fillna('')
+X = tfidf.fit_transform(text)
+
+# y
+ci = [list(i)[0] if len(i) == 1 else 0 for i in df.color_identity]
+le = preprocessing.LabelEncoder()
+y = le.fit_transform(ci)
+
+# Train-test split
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=111)
+```
+
+```{code-cell} ipython3
+plot_confusion_matrix(multiclass_model, X_test, y_test)  
+plt.show()
+```
+
+### Multilabel
+
+```{code-cell} ipython3
+from sklearn.feature_extraction.text import CountVectorizer
+from sklearn.svm import SVC
+from sklearn.multiclass import OneVsRestClassifier
+from sklearn.metrics import multilabel_confusion_matrix
+```
+
+```{code-cell} ipython3
+# Load the multilabel model
+multilabel_model = pickle.load(open("multilabel.sav", 'rb'))
+```
+
+Preprocessing:
+- min_df = 5, ignore rare words (appear in less than 5 documents)
+- max_df = 0.8, ignore common words (appear in more than 80% of documents)
+- Transform _color_identity_ into lowecase
+
+```{code-cell} ipython3
+# Instantiate OneVsRestClassifier
+multilabel_model = OneVsRestClassifier(SVC(kernel="linear"))
+
+# Train-test split
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=111)
+
+# Fit OneVsRestClassifier
+multilabel_model.fit(X_train, y_train)
+```
+
+```{code-cell} ipython3
+# Save the model using pickle
+pickle.dump(multilabel_model, open("multilabel.sav", 'wb'))
```

Comment:
Similar to the topic model, is this necessary? Should live in the source files, with this notebook only "reading" it, right? Did something not work from the pickle?